### PR TITLE
Escaping double quotes in EMBL/GenBank feature qualifiers (#1689)

### PR DESCRIPTION
**Pull request by alonalj**
Hi,

This pull request addresses issue #1689. 

FYI - this is my first contribution. I hope it's good to go, but please let me know if anything needs changing. I have added a unit test and verified it with build, install and test. Please let me know if anything needs changing. Thanks!

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file and understand that AppVeyor and
TravisCI will be used to confirm the Biopython unit tests and ``flake8`` style
checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

**Comment by alonalj** [comment]
#1689

**Comment by alonalj** [comment]
All tests seem to be getting "ok". I can't identify any obvious errors in the appveyor report except the network connectivity error... Is there something I'm missing?

**Comment by alonalj** [comment]
Managed to fix it - all should be good to go now...

File: Bio/GenBank/__init__.py [hunk 1]
```diff
@@ -1194,7 +1194,12 @@ def feature_qualifier(self, key, value):
             # otherwise just skip this key
             return
 
-        value = value.replace('"', '')
+        # keep GenBank's escaping requirement "" while removing single quotes "
+        # remove empty values (unique case of double quotes)
+        value = re.sub('^""$', '', value)
+        # remove single quotes unless for escaping purposes ""
+        value = value.replace('""', '~^').replace('"', '').replace('~^', '""')
```
Comment (peterjc) [inline Bio/GenBank/__init__.py]: Won't this break if ``value`` happens to have ``~^`` in it already? A simple if/else would be clearer I think.

**Comment by peterjc** [comment]
Hello @alonalj, thanks for working on this. 

For your test case, you have created this feature:

```
     source          1..154478
                     /organism="Arabidopsis thaliana"
                     /organelle="plastid:chloroplast"
                     /mol_type="genomic DNA"
                     /db_xref="taxon:3702"
                     /ecotype="Columbia"
                     /double_quotes_1="Arabidopsis ""thaliana"""
                     /double_quotes_2="Arabidopsis "" thaliana"
                     /double_quotes_3=""Arabidopsis thaliana""
```

Which you expect to be parsed as follows:

```python
        f.qualifiers['double_quotes_1'] = 'Arabidopsis ""thaliana""'
        f.qualifiers['double_quotes_2'] = 'Arabidopsis "" thaliana'
        f.qualifiers['double_quotes_3'] = '"Arabidopsis thaliana"'
```

I disagree.

By my reading, `` /double_quotes_1="Arabidopsis ""thaliana"""``removing the enclosing double-quote gives ``Arabidopsis ""thaliana""`` then undoing the escaping gives ``Arabidopsis "thaliana"``. But your test says you expect ``Arabidopsis ""thaliana""`` here.

Compare this to the NCBI provided example,

```
/note="This is an example of ""escaped"" quotation marks"
```

I think the parser value should be ``This is an example of "escaped" quotation marks`` here.

Also, it would be better to stick with NCBI standard keys, so using ``/note=...`` would be sensible but ``/double_quotes_1=...`` etc seems non-standard.

Also, could you use a smaller test file than ``NC_000932_qualif_esc.gb``? e.g. Start from AB070938 as used in many of the other tests - or better yet, pick an even shorter real record to work from.

**Comment by alonalj** [comment]
Thanks @peterjc . Agreed - I changed the read logic and created a smaller test file. I also added a write logic which addresses multiple cases, including partial escaping (see the test cases for examples). 
I didn't use a simple if-else option, as I'm not sure how the test examples could be easily addressed this way (unless you were referring to something other than: if "" replace with ", which won't be sufficient to cover all the tested options). If you had something else in mind and think it's worth changing to - let me know. Thanks.

File: Bio/GenBank/__init__.py [hunk 2]
```diff
@@ -1194,7 +1194,15 @@ def feature_qualifier(self, key, value):
...
 
-        value = value.replace('"', '')
+        # remove enclosing quotation marks
+        value = re.sub('^"|"$', '', value)
+
+        # Handle escaping requirements for quotation marks
+        # remove escaping if exists
+        value = re.sub('""', '"', value)
+        # add back " in case of improperly escaped empty quotes
+        value = re.sub(' " ', ' "" ', value)
```
Comment (peterjc) [inline Bio/GenBank/__init__.py]: What does this do and why?

File: Bio/GenBank/__init__.py [hunk 3]
```diff
@@ -1194,7 +1194,15 @@ def feature_qualifier(self, key, value):
             # otherwise just skip this key
             return
 
-        value = value.replace('"', '')
+        # remove enclosing quotation marks
+        value = re.sub('^"|"$', '', value)
+
+        # Handle escaping requirements for quotation marks
+        # remove escaping if exists
+        value = re.sub('""', '"', value)
```
Comment (peterjc) [inline Bio/GenBank/__init__.py]: Why not write this as:

```python
# Undo escaping, repeated double quotes -> one double quote
value = value.replace('""', '"')
```

File: Tests/GenBank/qualifier_escaping.gb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+FEATURES             Location/Qualifiers
+     source          1..182
+                     /organism="Mus musculus"
+                     /db_xref="taxon:10090"
+                     /chromosome="4"
+                     /map="4 42.6 cM"
+                     /note="Already ""escaped"""
+     Protein         1..182
+                     /product="interferon beta, fibroblast"
+                     /note="Not ""properly"" ""escaped"""
```
Comment (peterjc) [inline Tests/GenBank/qualifier_escaping.gb]: In what way is this example not properly escaped? It looks OK to me. Did you mean to have some single double-quote characters in the example?

File: Bio/GenBank/__init__.py (hunk 3 above)
Comment (peterjc) [reply Bio/GenBank/__init__.py]: Also, can you see a nice way to spot a double quote character which is not part of a two double quote pair? If so we can raise a parser warning about improperly escaped quotation marks.

**Comment by alonalj** [comment]
@peterjc  - I think I was operating under the wrong assumption that we should cover both cases where the user did not escape properly, as well as cases where the user did or even partially did (and this requires handling all sorts of unique cases). Maybe it will be much simpler if I just do the read and write logic assuming proper escaping by the user and we could add this to the documentation somewhere? If so - I'll just go ahead and replace the logic with the one-liners as you suggested, e.g. :
`# Undo escaping, repeated double quotes -> one double quote`
`value = value.replace('""', '"')`

**Comment by peterjc** [comment]
For the output, we can be strict and just escape everything.

For the parsing, we need to cope with real world input data, and do something reasonable. i.e. I would like a warning message if the GenBank input had things like this:

```
                     /note="The NCBI says double-quote characters like " should
                     be escaped as "" (two double-quotes), but here it was not"
```

**Comment by alonalj** [comment]
great, so just to verify - for the writing I'll leave only:
```
# Adhere to escaping guidelines
if type(value) == str and '"' in value:
    # add escaping, except where user already escaped
    value = value.replace('"', '""').replace('""""', '""')
```
(I've left the second replace to correct for cases where the user escaped, otherwise we'll have 4 on each side, but let me know if you have another idea in mind). 

And for the parsing - simply the one-liner mentioned earlier and a warning where we detected improper escaping. Using your text and warning examples from the genbank code, it would be of the following form:
```
if '"' in value and not re.search('"".*""', value):
    import warnings
    from Bio import BiopythonParserWarning
    warnings.warn('The NCBI states double-quote characters like should be escaped as ""(two double - 
    quotes), but here it was not ', BiopythonParserWarning)
```
(I could change the first line to if '"' in value and '""' not in value, but that won't capture "this"" case)

Any points for improvement or good to go?

**Comment by peterjc** [comment]
I was thinking just this for the writer code:

```
if type(value) == str:
    value = value.replace('"', '""')  # NCBI says escape " as "" in qualifier values
```

**Comment by peterjc** [comment]
For the parser, I was thinking something liek this:

```python
if '"' in value and '""' not in value:  # <-- this will miss some cases!
    import warnings
    from Bio import BiopythonParserWarning
    warnings.warn('The NCBI states double-quote characters like " should be escaped '
                  'as "" (two double-quotes), but here it was not: %r' % value,
                  BiopythonParserWarning)
```

The tricky bit would be that if-statement, probably a regular expression looking for a non-double quote, double-quote, non-double quote would work (i.e. something like ``[^"]"[^"]``); or start anchored double-quote, non-double quote; or end anchored non-double quote, double quote.

Or, rather than the ``[^X]`` syntax for anything other than ``X``, this might be done with a *Negative lookahead assertion*?

https://docs.python.org/3/howto/regex.html

**Comment by peterjc** [comment]
This is much improved @alonalj - glad to see that you could get the regular expression to work, and that you've got sensible test cases. Other than one tiny tweak to the test structure, I think the code this is ready to merge on the SeqIO side.

That's the good news - the bad news I thought of something else to do, sorry. I forgot about this until reviewing the test coverage - we have not considered the ``Bio.GenBank.read()`` parser, which is not ``SeqRecord`` based, this gives GenBank specific ``Records``. Here it may be reasonable to leave the quotes as is - at very least by doing nothing we round trip preserve them:

```
python3 -c "from Bio import GenBank; print(GenBank.read(open('qualifier_escaping_read.gb')))"
```

However, I think we should do the same unescaping and escaping here too.

For the input, you could move the new unescaping and warning to the Scanner code, which will thus fix both consumers (the SeqRecord consumer used by SeqIO, and the GenBank-specific Record consumer), i.e. move it here: https://github.com/biopython/biopython/blob/biopython-172/Bio/GenBank/Scanner.py#L400

For the output as GenBank format from a Record object this would be in this ``__str__`` method: https://github.com/biopython/biopython/blob/biopython-172/Bio/GenBank/Record.py#L633

In terms of documentation, a brief entry in the ``NEWS.rst`` file seems sensible to me - something like:

```
Double-quote characters in GenBank feature qualifier values are now escaped as
per the NCBI standard. Improperly escaped values trigger a warning on parsing.
```

**Comment by alonalj** [comment]
Made the changes (took a bit of extra tweaking due to the different handling of quotes with the GenBank parser as compared to the SeqIO one). Added read tests using GenBank.read() but not sure how to go about the writing part - all examples seem to be using SeqIO.

**Comment by peterjc** [comment]
Your comments like ``# Temporarily remove quotes to handle NCBI escaping`` made me wonder if this was the right thing to do after all - I was assuming the ``Record`` object's features didn't have the enclosing quotes, but evidently they were not removed.

```
$ python3 -c "from Bio import GenBank; print(GenBank.read(open('qualifier_escaping_read.gb')).features[2].qualifiers[0].value)"
"End not properly "escaped""
```

 i.e. If it was already returning the strings directly as-is, it makes sense for this parser to leave the quote escaping untouched?

i.e. We should go back to 43e0ad8 and just add more specific a ``NEWS.rst`` entry:

```
Double-quote characters in GenBank feature qualifier values in ``Bio.SeqIO``
are now escaped as per the NCBI standard. Improperly escaped values trigger a
warning on parsing.
```

If you agree, say so, and you can leave me to handle this (since I made this extra work).

**Comment by peterjc** [comment]
See also #1817, should we deprecate Bio.GenBank.Record based GenBank parser?

**Comment by alonalj** [comment]
No worries :) - added a new pull request for the previous version, with the addition of splitting the test into read and write see  #1818
As for deprecating - it seems fitting as multiple places state it is not recommended to use it...

**Comment by peterjc** [comment]
Closed in favour of #1818, thanks.

**Pull request closed without merging**